spot-vaults/contracts/_strategies/SpotAppraiser.sol

```diff
@@ -0,0 +1,257 @@
+    // Constructor
+
+    /// @notice Contract constructor.
+    /// @param ampl Address of the AMPL token.
+    /// @param spot Address of the SPOT token.
+    /// @param usdOracle Address of the USD token market price oracle token.
+    /// @param amplOracle Address of the AMPL market price oracle token.
+    constructor(
+        IAMPL ampl,
+        IPerpetualTranche spot,
+        IChainlinkOracle usdOracle,
+        IChainlinkOracle amplOracle
+    ) Ownable() {
+        AMPL = ampl;
+        SPOT = spot;
+        AMPL_ORACLE = amplOracle;
+        USD_ORACLE = usdOracle;
+        UNIT_AMPL = 10 ** IERC20Metadata(address(AMPL)).decimals();
+
+        tolAMPLPriceDeviationPercs = Range({
+            lower: (ONE * 8) / 10, // 0.8
+            upper: (ONE * 7) / 4 // 1.75
+        });
+        minSPOTDR = (ONE * 8) / 10; // 0.8
+        minSeniorCDR = (ONE * 5) / 4; // 125%
+    }
+
+    //--------------------------------------------------------------------------
+    // Owner only methods
+
+    /// @notice Controls the tolerable AMPL price volatility outside which the pricing strategy is considered unreliable.
+    /// @param tolAMPLPriceDeviationPercs_ The lower and upper percentage deviation of AMPL price.
+    function updateAllowedPriceDeviationPercs(
+        Range memory tolAMPLPriceDeviationPercs_
+    ) external onlyOwner {
+        if (
+            tolAMPLPriceDeviationPercs_.lower > ONE ||
+            tolAMPLPriceDeviationPercs_.upper < ONE
+        ) {
+            revert InvalidPerc();
+        }
+        tolAMPLPriceDeviationPercs = tolAMPLPriceDeviationPercs_;
+    }
+
+    /// @notice Controls the minimum `deviationRatio` ratio of SPOT below which SPOT is considered unhealthy.
+    /// @param minSPOTDR_ The minimum SPOT `deviationRatio`.
+    function updateMinSPOTDR(uint256 minSPOTDR_) external onlyOwner {
+        minSPOTDR = minSPOTDR_;
+    }
+
+    /// @notice Controls the minimum CDR of SPOT's senior tranche below which SPOT is considered unhealthy.
+    /// @param minSeniorCDR_ The minimum senior tranche CDR.
+    function updateMinPerpCollateralCDR(uint256 minSeniorCDR_) external onlyOwner {
+        if (minSeniorCDR_ < ONE) {
+            revert InvalidSeniorCDRBound();
+        }
+        minSeniorCDR = minSeniorCDR_;
+    }
+
+    //--------------------------------------------------------------------------
+    // External methods
+
+    /// @return p The price of the usd token in dollars.
+    /// @return v True if the price is valid and can be used by downstream consumers.
+    function usdPrice() external view override returns (uint256, bool) {
+        (uint256 p, bool v) = _getCLOracleData(USD_ORACLE);
+        // If the market price of the USD coin fallen too much below 1$,
+        // it's an indication of some systemic issue with the USD token
+        // and thus its price should be considered unreliable.
+        return (ONE, (v && p > USD_LOWER_BOUND));
```

Comment:
This logic is saying that if the market price of the dollar coin is less than the lower bound (.99 now), then the oracle report is bad. But, it could be a good report representing the real market value of the dollar coin.

Whether we want to operate that far outside the bands of normality is still a sane question to ask... but that's application specific rather than oracle specific. The oracle's job is to deliver data. WDYT, should we move this check into the business logic?

Reply:
The approach here is at times of volatility (or low confidence on the pricing function, I think of both interchangeably), the bill-broker stops all trading activity. Similar to an mm pull liquidity and waiting out the storm ..

If the price of the dollar coin deviates a lot from its 1$ peg, its an indication that we now have lower confidence in the the pricing function we use (ie 1 dollar coin = 1 dollar).